### PR TITLE
Implement sys_rsx_context_free

### PR DESCRIPTION
**Pull request by elad335**
Used by VSH.

### CODE REVIEW

File: rpcs3/Emu/Cell/lv2/sys_rsx.cpp [hunk 1]
```diff
@@ -140,7 +140,9 @@ error_code sys_rsx_memory_free(cpu_thread& cpu, u32 mem_handle)
 		return CELL_ENOMEM;
 	}
 
-	if (rsx::get_current_renderer()->dma_address)
+	const auto render = rsx::get_current_renderer();
```
Comment (kd-11) [inline rpcs3/Emu/Cell/lv2/sys_rsx.cpp:143]: What's going on here? This variable is never used again.

File: rpcs3/Emu/Cell/lv2/sys_rsx.cpp [hunk 2]
```diff
@@ -276,13 +278,21 @@ error_code sys_rsx_context_free(cpu_thread& cpu, u32 context_id)
 
 	const auto render = rsx::get_current_renderer();
 
-	std::scoped_lock lock(render->sys_rsx_mtx);
+	std::unique_lock lock(render->sys_rsx_mtx);
```
Comment (kd-11) [inline rpcs3/Emu/Cell/lv2/sys_rsx.cpp:281]: sys_rsx.todo above should be sys_rsx.warning since it is now implemented.

Comment (elad335) [reply rpcs3/Emu/Cell/lv2/sys_rsx.cpp:281]: Nah, it's a hacked implementation, I'll leave it TODO. Also this is a rare syscall that you want to be always noticed warning doesn't make sense here.

File: rpcs3/Emu/Cell/lv2/sys_rsx.cpp (hunk 1 above)
Comment (elad335) [reply rpcs3/Emu/Cell/lv2/sys_rsx.cpp]: Fixed.

**Review by kd-11 — APPROVED** [review]

**Comment by Vestrel** [comment]
Now it crashes during thread cleanup

**Comment by Megamouse** [comment]
Is that a blocker for this PR or is it unrelated?

**Comment by elad335** [comment]
Dunno, depends on where it crashes.

**Comment by Vestrel** [comment]
During execution of cleanup() in cpu_thread::operator() of rsx::thread

**Comment by elad335** [comment]
Visual studio pic of crash should suffice.

**Comment by Vestrel** [comment]
image

**Comment by elad335** [comment]
It's not RSX related then.

**Comment by kd-11** [comment]
I'll try to repro and provide a fixup.

**Comment by elad335** [comment]
I mean, this is a PPU crash.

**Comment by kd-11** [comment]
It's caused by the PR regardless.
It happens because you called the cleanup operator on rsx thread from an external thread which leads to a fatal error here: https://github.com/RPCS3/rpcs3/blob/master/rpcs3/Emu/CPU/CPUThread.cpp#L360
Simply remove the `(*render)()` call, it is unnecessary here imo. By the way, RSX has already exited (destructor is already executed and all) by the time context free is even being called.
sys_rsx_context free is not really relevant at the moment whilst driver state and hw state are still merged together. Just dealloc is enough for now.

**Comment by kd-11** [comment]
P.S
There is still an unrelated issue on exit that is truly PPU related, where sys_uart_receive crashes after system shutdown but it has a completely different stacktrace and all so I feel we can ignore that particular one.

**Comment by elad335** [comment]
Oh, I see, I forgot that rsx::thread is not named_thread wotgout casting and operator() has different purposes between them.

**Comment by elad335** [comment]
@kd-11 That's not what I meant.

**Comment by kd-11** [comment]
Ok, I'll leave it up then, its not critical.

**Comment by brian218** [comment]
Why were the modifications of sys_rsx.cpp eliminated in this PR?

**Comment by kd-11** [comment]
Seems broken. Should be closed.

**Pull request closed without merging**